src/usb.c

```diff
@@ -22,66 +23,70 @@ struct usb_drd_regs {
     uintptr_t atc;
 };
 
-static const struct {
-    const char *dart_path;
-    const char *dart_mapper_path;
-    const char *atc_path;
-    const char *drd_path;
-    const char *hpm_path;
-} usb_drd_paths[2] = {
-    {
-        .dart_path = "/arm-io/dart-usb0",
-        .dart_mapper_path = "/arm-io/dart-usb0/mapper-usb0",
-        .atc_path = "/arm-io/atc-phy0",
-        .drd_path = "/arm-io/usb-drd0",
-        .hpm_path = "/arm-io/i2c0/hpmBusManager/hpm0",
-    },
-    {
-        .dart_path = "/arm-io/dart-usb1",
-        .dart_mapper_path = "/arm-io/dart-usb1/mapper-usb1",
-        .atc_path = "/arm-io/atc-phy1",
-        .drd_path = "/arm-io/usb-drd1",
-        .hpm_path = "/arm-io/i2c0/hpmBusManager/hpm1",
-    },
-};
+#define FMT_DART_PATH        "/arm-io/dart-usb%u"
+#define FMT_DART_MAPPER_PATH "/arm-io/dart-usb%u/mapper-usb%u"
+#define FMT_ATC_PATH         "/arm-io/atc-phy%u"
+#define FMT_DRD_PATH         "/arm-io/usb-drd%u"
+#define FMT_HPM_PATH         "/arm-io/i2c0/hpmBusManager/hpm%u"
 
 static tps6598x_irq_state_t tps6598x_irq_state[USB_INSTANCES];
 static bool usb_is_initialized = false;
 struct usb_drd_regs usb_regs[USB_INSTANCES];
 
-static dart_dev_t *usb_dart_init(const char *path, const char *mapper_path)
+static int snprintf(char *buffer, size_t size, const char *fmt, ...)
+{
+    va_list args;
+    int len;
+
+    va_start(args, fmt);
+    len = vsnprintf(buffer, size, fmt, args);
+    va_end(args);
+    if (len >= 0 && (size_t)len >= size)
+        printf("usb: Error praparing ADT path '%s'", fmt);
+
+    return len;
+}
+
+static dart_dev_t *usb_dart_init(u32 idx)
 {
     int mapper_offset;
+    char path[sizeof(FMT_DART_MAPPER_PATH)];
```

Comment:
Might be good adding a comment that this is good up to 100 instances, since that isn't immediately obvious unless you think about the width of the format specifier.

Reply:
comment and `#error` added